### ISSUE TITLE
Windows OS - UI appears broken

### ISSUE THREAD

**Issue by UberWaffe** [Pulsar4xDevs/Pulsar4x#225]
When compiling the code (used Visual Studio 2017) for Windows_x64 (using Windows 8), the UI appears broken when playing the game.
(Used commit c3c48864caaefca03e419e017738cf2628a101ca)

Orbits do not show.
Entities (planets, stars, etc.) do not show.
Buttons appears to show as garbled icons.

![Pulsar4x_Windows_UI_02](https://user-images.githubusercontent.com/2726355/59694978-c2049b00-91e9-11e9-8d9f-f856f69db3c7.PNG)
![Pulsar4x_Windows_UI_01](https://user-images.githubusercontent.com/2726355/59694980-c2049b00-91e9-11e9-9005-3ed52b431208.PNG)

Opening the cargo on a ship has the ship name overlap the new popup window.

![Pulsar4x_Windows_UI_03](https://user-images.githubusercontent.com/2726355/59695129-0132ec00-91ea-11e9-97d5-1400e8dced3b.PNG)

**Comment by se5a** [comment]
I'm going to hazard a guess that this is an SDL problem. 
After you start the game, hit esc, click "Options" then check the "Show ImgDebug" checkbox.
These two lines here should print some text to the ui if SDL failed:
https://github.com/Pulsar4xDevs/Pulsar4x/blob/Master/Pulsar4X/Pulsar4X.ImGuiNetUI/Program.cs#L178

Though the fact that they're displaying but garbled it might be a bit of a long shot that it picks anything up. 
I'll try find my netbook and hopefully it'll run it enoguh to see whats going on my end. 
This side of things I'm pretty weak-source at though, most the SDL integration was cargo culted.

**Comment by se5a** [comment]
So I managed to get it running in VirtualBox, there's something odd with that, but that's another issue. 

yeah Show ImgDebug is showing that it doesn't like something.
I pushed a change to use Path.Combine to ensure there were no issues with / vs \ when trying to load the bitmaps, but the problem is deeper than that. 
it apears that  `rendererPtr = SDL.SDL_GetRenderer(windowPtr);`  [in line 67 of the GlobalUIState](https://github.com/Pulsar4xDevs/Pulsar4x/blob/Master/Pulsar4X/Pulsar4X.ImGuiNetUI/GlobalUIState.cs#L67) is returning zeros. 
SDL.GetError() doesn't show anything untoward at this point either. will have to do more research.

**Comment by se5a** [comment]
@UberWaffe  Try commit https://github.com/Pulsar4xDevs/Pulsar4x/commit/9bfd05b25fe6cd7777d833b05c363d03b717f4b5

I don't know if this will work or if it will crash. 
I don't fully understand how SDL is setup and integrated with IMGUI, I don't see an SDL Renderer being created anywhere, so how linux gets one for free... IDK. I've thrown a check in there and it'll try create one if one does not exist.  
Running in VirtualBox I was getting a System.OverflowException: 'Arithmetic operation resulted in an overflow' [here](https://github.com/Pulsar4xDevs/Pulsar4x/blob/Master/Pulsar4X/Pulsar4X.ImGuiNetUI/IMGUISDL/ImGuiSDL2CSHelper.cs#L157) which doesn't make a whole lot of sense too me. 
see if you get a similar result.

**Comment by UberWaffe** [comment]
@se5a Awesome, will give it a shot and let you know the outcome.

**Comment by UberWaffe** [comment]
@se5a The error "System.OverflowException: 'Arithmetic operation resulted in an overflow" you mentioned seems to have to do with 64-bit versus 32-bit operations.
https://stackoverflow.com/questions/18173211/how-to-convert-intptr-to-int
Basically in a 64-bit environment it is equivalent to long, not int.

However, even casting to long doesn't fix the graphical issues. I'm still mucking about with it.
Note: I am using 64-bit operating systems.

Sidenote: I also updated the ImGUI.Net nuget package to a more recent version, and it didn't make a difference. Got the same graphical anomalies.

**Comment by se5a** [comment]
Wonder if it's trying to use a 32 bit version of SDL or some other library for windows then. I'll have a closer look tomorrow.

**Comment by se5a** [comment]
the buttons and images are all the font atlas, which I have a fuzzy recollection is what imgui defaults to if it's having a problem with the image.

**Comment by se5a** [comment]
I've changed it to store the Fonttexture as an IntPtr.
that FontTextureID is just the font atlas. 
Our problem lies with `var sdlid = SDL.SDL_GL_BindTexture(pcmd.TextureId, out w, out h);` [ here: ](https://github.com/Pulsar4xDevs/Pulsar4x/blob/Master/Pulsar4X/Pulsar4X.ImGuiNetUI/IMGUISDL/ImGuiSDL2CSHelper.cs#L173)
If you put a break just after that, you can inspect the SDL_GetError() string. it'll show "That operation is not supported".
Not sure yet what's causing that. it's also a worry that it's not displaying then SDL_RenderDrawLine() calls either. though those don't seem to be throwing an SDL error. 

I'll have to think about collecting these errors and displaying them somewhere. 

Also, in my VM the text is fuzzy, looks like yours isn't? hopefully that's just a result of it running in a VM.

**Comment by UberWaffe** [comment]
I've not had issues with fuzzy text.

Sidenote: I've played around with using newer versions of the SDL-CS and ImGui libraries, but to no avail.

**Comment by se5a** [comment]
yeah I'm pretty confident that it's not ImGui, I had trying out a newer version of SDL-CS on my list of things to try but hadn't yet, so I can cross that one off. 
interestingly I found that calling SLD_GetWindowSurface seems to create and SDL_Renderer. at least, on linux anyway. I've kinda hit a brick wall of things to poke in that direction and have asked on the SDL forums for some ideas of how to chase down why it's not playing the game. 
I'm still suspicious of the SDL_Renderer, I think that may be where the problem is (the GL_BindTexture uses it internaly so it's either that or the texture itself) 
so much for my "if it works in linux it should be reasonably easy to get it working in windows" idea.

**Comment by UberWaffe** [comment]
The little bit of reading up I did also indicated in the direction of SDL_Renderer.
Haven't made any progress yet.

**Comment by se5a** [comment]
Fixed it, get the latest and confirm. 

problem was that on windows it defaults to Direct3d so GL calls fail. just needed to tell it to use openGL. 
Confirm it works your end and close this if it does.

One thing I noticed in the VM is that on windows it doesn't seem to be adjusting the position of the names if they're close to each other and stacking them ontop of each other. but that's another issue, I want to do all that differently anyway.
